### PR TITLE
Fix #45: add spider for budget receipts data

### LINKED ISSUE #45: Add spider for budget receipts

### PR DESCRIPTION
**Pull request by curioswati**

### CODE REVIEW

**Comment by curioswati** [comment]
please only review till the commit: `e419cf8`, the latest is just a rebase from another branch.

File: scraper/scraper/spiders/budget_base.py [hunk 1]
```diff
@@ -1,43 +1,71 @@
...
+
     def make_dataset_request(self, params):
         '''
         Construct and yield a scrapy request for a dataset.
         '''
         # generate a filepath to store the dataset in.
         filename = '{}_{}.csv'.format(self.name, self.date)
 
-        filepath = os.path.join(DATASET_PATH, filename)
+        filepath = os.path.join(BUDGET_DATA_PATH, self.name.split('_')[-1], filename)
```
Comment (gggodhwani) [inline scraper/scraper/spiders/budget_base.py]: Make the filepath name much more verbose

File: scraper/scraper/spiders/budget_base.py [hunk 2]
```diff
@@ -146,13 +175,138 @@ def parse_dataset(self, response):
...
+    Base spider for HP budget receipts.
+    '''
+    def __init__(self, *args, **kwargs):
+        super(ReceiptBaseSpider, self).__init__(*args, **kwargs)
+        if not hasattr(self, 'date'):
+            raise CloseSpider('No date given!')
+
+        self.date = kwargs.pop('date')
+
+        if not re.match(r'\d{2}/\d{2}/\d{4}', self.date):
```
Comment (gggodhwani) [inline scraper/scraper/spiders/budget_base.py]: Make a variable for date regex

File: scraper/scraper/spiders/budget_base.py [hunk 3]
```diff
@@ -146,13 +175,138 @@ def parse_dataset(self, response):
...
+            raise CloseSpider('Wrong date format, please use dd/mm/yyyy!')
+
+    @classmethod
+    def from_crawler(cls, crawler, *args, **kwargs):
+        spider = super(ReceiptBaseSpider, cls).from_crawler(crawler, *args, **kwargs)
+        crawler.signals.connect(spider.spider_closed, signal=signals.spider_closed)  #pylint: disable=no-member
```
Comment (gggodhwani) [inline scraper/scraper/spiders/budget_base.py]: Add reference to small guideline/doc on Scrapy signals

File: scraper/scraper/spiders/budget_base.py [hunk 4]
```diff
@@ -146,13 +175,138 @@ def parse_dataset(self, response):
...
+            heads = ['Maj-Smj-Min-Smn [Sub Minor Head]', 'Total Receipt']
+            csv_writer = csv.writer(self.dataset_file, delimiter=',')
+            csv_writer.writerow(heads)
+
+            for public_head_bool in ['0', '1']:
+                params = {
+                    'date': self.date,
+                    'unit': self.unit,
+                    'public_head_bool': public_head_bool,
+                    'major_code': '0000',  # we use all major heads
```
Comment (gggodhwani) [inline scraper/scraper/spiders/budget_base.py]: 0000 is code for "All Major Heads"

File: scraper/scraper/spiders/budget_exp_receipts.py [hunk 5]
```diff
@@ -12,4 +18,15 @@ class BudgetExpendituresSpider(budget_base.BudgetBaseSpider):
 
     query_index = 9
 
-    unit = '.00001'
\ No newline at end of file
+    unit = '.00001'
```
Comment (gggodhwani) [inline scraper/scraper/spiders/budget_exp_receipts.py]: Just check if the unit, and local minima. Plus rename to explain the decimals

**Review by gggodhwani — CHANGES_REQUESTED** [review]

**Review by gggodhwani — APPROVED** [review]
:+1:

**Pull request merged**